Add SSO support

skovati opened the pull request:
closes https://github.com/NASA-AMMOS/aerie/issues/1100

This PR significantly reworks the `hooks.server.ts` logic to support SSO in addition to the `/login` page. Using a new environment variable `PUBLIC_AUTH_SSO_ENABLED`, the top level hook auth flow is now as follows:

```javascript
if (env.PUBLIC_AUTH_SSO_ENABLED === 'true') {
   return await handleSSOAuth({ event, resolve });
} else {
   return await handleJWTAuth({ event, resolve });
}
```

### `handleJWTAuth`
`handleJWTAuth` is just a refactored version of the existing auth flow:
 - validate + populate `event.locals.user` using existing JWT cookie (by hitting Gateway's `/auth/session`)
 - if JWT was invalid / nonexistent, redirect to `/login`

### `handleSSOAuth`
`handleSSOAuth` is a new flow that supports SSO auth by forwarding cookies to new Gateway endpoints, which will check SSO token validity (e.g. CAM token) and return JWTs or redirections accordingly:
 - forward cookies to `/auth/validateSSO`, which will run an auth provider specific SSO validation
 - if above validation was successful, Gateway returns an Aerie JWT, which we store in cookies and use in `Auth:` headers as usual.
 - else, the SSO token was invalid or DNE, Gateway returns the URL of the auth providers login page (e.g. CAM-UI), which we redirect the user to.

## Logout
The UI also switches on this `PUBLIC_AUTH_SSO_ENABLED` env var during logout, where it'll either call the Gateway's `/auth/logoutSSO` (which invalidates the SSO token using the auth provider), or just clear the JWT and redirect to `/`.

On the next page load, the UI will run through the auth flow in `hooks.server.ts`, which will correctly redirect us to either `/login` or the SSO login page, depending on `PUBLIC_AUTH_SSO_ENABLED`.

This design allows the UI to remain auth provider agnostic. The Gateway can then be extended to support other auth providers (e.g. https://aws.amazon.com/cognito).

The corresponding Gateway PR (https://github.com/NASA-AMMOS/aerie-gateway/pull/51) defines these new auth endpoints, as well as lists the testing steps and possible env var configurations.

skovati committed:
add new `reqGateway` function that passes cookies

skovati committed:
Rework auth hook logic to support SSO

duranb reviewed (changes requested):
I set `AUTH_TYPE` to "none", but it still redirects me to the CAM login. Also, after logging in to CAM, I wasn't redirected back to aerie-ui. I think whatever is doing the redirect is not including the `goto` return url to CAM

skovati committed:
Add support for new gateway auth endpoints.

This commit also removes the auth redirection logic from the UI, since
this is handled by the gateway instead, which will return a redirect if
deemed necessary.

skovati committed:
fix lint errors

skovati committed:
update tests for new login flow

duranb commented on src/utilities/login.ts:
```diff
@@ -1,22 +1,16 @@
   if (browser) {
     await fetch(`${base}/auth/logout`, { method: 'POST' });
-    await goto(`${base}/login${reason ? '?reason=' + reason : ''}`, { invalidateAll: true });
+    await goto(`${base}/`);
```
you can also take out the trailing `/`

duranb committed:
add ability to start local UI with https

skovati committed:
add referrer to validation requests

duranb committed:
add ability to specify local host domain

skovati committed:
fix redirection logic

skovati committed:
decode URI encoded cookies

skovati committed:
run prettier

skovati committed:
throw redirect instead of returning

prevents `user` being typed as possibly undefined, which svelte-check didn't like

skovati committed:
remove `PUBLIC_LOGIN_PAGE`

"NoAuthAdapter" now fulfills the use case where no auth is desired

skovati committed:
add example env vars for local HTTPS + domain dev

duranb commented on .env:
```diff
@@ -5,3 +5,5 @@ PUBLIC_HASURA_CLIENT_URL=http://localhost:8080/v1/graphql
 PUBLIC_HASURA_SERVER_URL=http://localhost:8080/v1/graphql
 PUBLIC_HASURA_WEB_SOCKET_URL=ws://localhost:8080/v1/graphql
 PUBLIC_LOGIN_PAGE=enabled
```
can you also remove `PUBLIC_LOGIN_PAGE` from here as well, please?

skovati committed:
add new `PUBLIC_AUTH_TYPE` env var

skovati committed:
switch auth flow based on new env var

skovati committed:
run prettier

skovati commented on docs/ENVIRONMENT.md:
```diff
@@ -11,4 +11,3 @@ This document provides detailed information about environment variables for Aeri
 | `PUBLIC_HASURA_CLIENT_URL`       | Url of Hasura as called from the client (i.e. web browser)                                                | `string` | http://localhost:8080/v1/graphql |
 | `PUBLIC_HASURA_SERVER_URL`       | Url of Hasura as called from the server (i.e. Node.js container)                                          | `string` | http://localhost:8080/v1/graphql |
 | `PUBLIC_HASURA_WEB_SOCKET_URL`   | Url of Hasura called to establish a web-socket connection from the client                                 | `string` | ws://localhost:8080/v1/graphql   |
```
Add new env var description

skovati commented on docker-compose-test.yml:
```diff
@@ -8,7 +8,8 @@ services:
     environment:
-      AUTH_TYPE: none
+      AUTH_TYPE: default
+      AUTH_UI_URL: 'http://localhost/login'
```
Add new env var

skovati commented on src/hooks.server.ts:
```diff
@@ -66,3 +26,122 @@ export const handle: Handle = async ({ event, resolve }) => {
+
+  const allowedRoles = decodedToken['https://hasura.io/jwt/claims']['x-hasura-allowed-roles'];
+  const defaultRole = decodedToken['https://hasura.io/jwt/claims']['x-hasura-default-role'];
+  activeRole ??= defaultRole;
```
dont do this

skovati commented on src/utilities/login.ts:
```diff
@@ -1,22 +1,16 @@
-export async function logout(reason?: string) {
+export async function logout() {
   if (browser) {
     await fetch(`${base}/auth/logout`, { method: 'POST' });
```
add sso check

skovati committed:
change auth type env var to boolean

skovati committed:
document new sso env var

skovati committed:
refactor nullish assign

skovati committed:
fix test env vars

skovati committed:
restore feature parity with logout reason

skovati committed:
add error handling to cookie parsing

skovati committed:
fix role switching with SSO flow

skovati committed:
add env var mock to login tests

duranb commented on src/hooks.server.ts:
```diff
@@ -66,3 +26,130 @@ export const handle: Handle = async ({ event, resolve }) => {
+  const userBuffer = Buffer.from(userCookie ?? '', 'base64');
+  const userStr = userBuffer.toString('utf-8');
+
+  try {
```
Should this try catch go instead in `computeRolesFromJWT`? It looks like that function is already returning a `null` so I figured you can just return the output from `computeRolesFromJWT` here.

skovati replied:
I added this try catch so that the UI server wouldn't crash if a malformed `user` cookie got sent, i.e. `JSON.parse()` fails. The `computeRolesFromJWT` call is only within the try catch so it shares scope with `const baseUser`. 🤔

skovati replied:
I agree this is a little awkward to read, reads like it's trying to catch an exception from `computeRolesFromJWT`

duranb replied:
Ohhh. My bad. I didn't see the JSON.parse...

skovati committed:
fix redirection loop by replacing `isDataRequest` conditional

skovati committed:
fix redirect loop

skovati committed:
fix logout cookie setting race condition

AaronPlave commented:
Tracking the separate issue that we talked about here https://github.com/NASA-AMMOS/aerie-ui/issues/1106 @skovati @duranb
